File: app/static/js/local/buildingManagement.js
// Javascipt file for Building Management
// console.log("Javascript loaded!")

var rIDGlobal = "";
var lastRoomID = null;

function setRoomId(rid){
    rIDGlobal=parseInt(rid);
}
function getRoomId(){
    return rIDGlobal;
}

function setRoomInfo(roomID, button){
    // '''For populating panel onClick of Edit button. updateHTML is called'''
    //Sets room ID based on what room (row) Edit button was clicked
    // A new room was selected; show roomDetails

    if (roomID != lastRoomID) {
      setRoomId(roomID);
      movePanel();
      // console.log("changing RID from " + lastRoomID + " to " + roomID)
      lastRoomID = roomID;
      getRoomDetails();
      $("#roomDetails").collapse('show');
    } else {
      $("#roomDetails").collapse('toggle');
      // lastRoomID = null;
    }

}

function movePanel() {
    //Takes rID to ensure correct room per row
    //Called in setRoomInfo
    var targetDiv = document.getElementById("hiddenRow_"+getRoomId());// hidden row where content will be placed
    var sourceDiv = document.getElementById("roomDetails");// content to be placed in targetDiv
    $(targetDiv).html($(sourceDiv)); // moves modal content into current row
    // $(sourceDiv).collapse('show');
}

function updateHtml(response) {
    //Updates the HTML in panel.
    //Called in AJAX of setRoomInfo()
    // console.log("Starting updateHTML");

    $('#roomNumber').text((response['building'] + " " + response['number']).toString()).data('text');
    var my_div = document.getElementById('roomCapacity');
    my_div.value = response['capacity'];
    var my_div = document.getElementById('roomType');
    my_div.value = response['type'];
    var my_div = document.getElementById('specializedEq');
    my_div.value = response['specializedEq'];
    var my_div = document.getElementById('specialFeatures')
    my_div.value = response['specialFeatures'];
    var my_div = document.getElementById('movableFurniture');
    // console.log("moveble: ", my_div.value)
    // my_div.removeAttribute("checked"); //It was getting stuck
    if (response['movableFurniture']) {
        my_div.setAttribute("checked", "checked");
    }

    // Update the selectpickers with values from DB
    aa = $("#accessibility #audioAcc");
    va = $("#accessibility #visualAcc");
    pa = $("#accessibility #physicalAcc");
    aa.selectpicker();
    va.selectpicker();
    pa.selectpicker();
    if (response['audioAcc'] != null) {
      $("#audioAcc").selectpicker('val', response['audioAcc']);
    } else {
      $("#audioAcc").selectpicker('val', 0);
    };
    if (response['visualAcc'] != null) {
      $("#visualAcc").selectpicker('val', response['visualAcc']);
    } else {
      $("#visualAcc").selectpicker('val', 0);
    };
    if (response['physicalAcc'] != null) {
      $("#physicalAcc").selectpicker('val', response['physicalAcc']);
    } else {
      $("#physicalAcc").selectpicker('val', 0);
    };
    $("#audioAcc").selectpicker('refresh');
    $("#visualAcc").selectpicker('refresh');
    $("#physicalAcc").selectpicker('refresh');
};

function getRoomDetails() {
  // console.log(this.id + " was shown")
  if (rIDGlobal > 0){
           var url = '/getRoomData/'+rIDGlobal;
           $.ajax({
                  url: url,
                  dataType: 'json',//Corresponds with controller json dumps
                  type: "GET",
                  success: function(response){
                    // console.log(response);
                      if (response["success"] != 0) {//If successful
                          updateHtml(response);//Update the panel with the data
                          doDropzoneSetup();
                      }
                  },
                  error: function(error) {
                      console.log("Error" + error);
                  }
              });
  }
};

function doDropzoneSetup() {
  if (typeof dz != "undefined") {
    dz.remove();
    newdz = "<div id='myId' class='dropzone'></div>";
    $("#imageUploaderDiv").append(newdz);
  };
  dz = $("div#myId").dropzone({ url: "/imageUpload/" + getRoomId(),
                                renameFile: function (file) {
                                  return "room_" + getRoomId() + "_" + new Date().getTime() + '.' + file.name.split(".").pop();
                                },
                                addRemoveLinks: true,
                                parallelUploads: 1,
                                thumbnailWidth:"120",
                                thumbnailHeight:"120",
                                acceptedFiles: "image/*",
                                removedfile: function (file) {
                                  // console.log("Remove from server");
                                  $.ajax({
                                     type: "POST",
                                        url: "/removeImage",
                                        data: {"rid": getRoomId(),
                                               "file": file.name
                                              },   //Dictionary pass
                                        dataType: 'json',
                                        success: function(response){
                                          // console.log(response)
                                          file.previewElement.remove();
                                        },
                                        error: function(error){
                                          console.log(error);
                                        }
                                  });
                                },
                                init: function () {
                                  let myDropzone = this;
                                  $.ajax({
                                     type: "GET",
                                        url: "/getImages/" + getRoomId(),
                                        dataType: 'json',
                                        success: function(response){
                                          // console.log(response)
                                          for (i = 0; i < response.length; i++) {
                                            filename = response[i]["name"]
                                            fileExt = filename.split(".").pop()
                                            fileSize = response[i]["size"]
                                            var mockFile = {
                                                name: filename,
                                                type: 'image/' + fileExt,
                                                size: fileSize
                                            };

                                            myDropzone.displayExistingFile(mockFile, "/static/images/" + response[i]["name"]);
                                          }
                                        },

                                        error: function(error){
                                          console.log(error);
                                        }
                                  });
                                }});
  // console.log(getRoomId());

};

function saveChanges(roomID){
    // KEEPING FOR POSTERITIES SAKE - SH

    //Sets up datetime, passes all room attributes into a dictionary for ajax,then posts data to DB and reloads the page

    // //Datetime setup
    // var dateTime = new Date();
    // var currHour = dateTime.getHours();
    // if (currHour < 12) //AM/PM setup
    //    {
    //    a_p = "AM";
    //    }
    // else
    //    {
    //    a_p = "PM";
    //    }
    // if (currHour == 0)
    //    {
    //    currHour = 12;
    //    }
    // if (currHour > 12)
    //    {
    //    currHour = currHour - 12;
    //    }
    //
    // var currMin = dateTime.getMinutes();
    // currMin = currMin + "";
    // if (currMin.length == 1) //Getting JS to not do single digit minutes
    // {
    //     currMin = "0" + currMin;
    // }
    // var savedDateTime=(dateTime.toDateString()+",  "+ currHour + ":" +currMin + " " + a_p); //Concatenation of all date elements into one var for passing into dictionary
    //End datetime setup

    //Begin dictionary pass for Ajax
    var roomDetails = {}//For passing into Ajax data field (multiple attributes to pass)
    roomDetails["roomCapacity"] = document.getElementById('roomCapacity').value;
    roomDetails["roomType"] = document.getElementById('roomType').value;
    roomDetails["specializedEq"] = document.getElementById('specializedEq').value;
    roomDetails["specialFeatures"] = document.getElementById('specialFeatures').value;
    roomDetails["movableFurniture"] = document.getElementById('movableFurniture').checked;
    roomDetails["visualAcc"] = $('#visualAcc option:selected').text();
    roomDetails["audioAcc"] = $('#audioAcc option:selected').text();
    roomDetails["physicalAcc"] = $('#physicalAcc option:selected').text();

    var url = '/saveChanges/'+getRoomId();
    $.ajax({
       type: "POST",
          url: url,
          data: roomDetails,                              //Dictionary pass
          dataType: 'json',
          success: function(response){
                  window.location = "/buildingManagement" //Refresh page
          },
          error: function(error){
              console.log("ERROR" + error)
              window.location.assign("/buildingManagement")
          }
    });


}


  /*sets the radio checkbox for education tech for each room*/
function education_detail(response){
    console.log("Starting ed tech filling");
    var my_div = document.getElementById('projectors');
    my_div.value = response['projector'];
    var my_div = document.getElementById('smartboards');
    my_div.value = response['smartboards'];
    var my_div = document.getElementById('instructor_computers');
    my_div.value = response['instructor_computers'];
    var my_div = document.getElementById('podiums');
    my_div.value = response['podium'];
    var my_div = document.getElementById('student_workspace');
    my_div.value = response['student_workspace'];
    var my_div = document.getElementById('chalkboards');
    my_div.value = response['chalkboards'];
     var my_div = document.getElementById('whiteboards');
    my_div.value = response['whiteboards'];

    var my_div1 = $("#vhs");
    if (response['vhs']) {
	my_div1.prop('checked', true);
    } else {
	my_div1.prop('checked', false);

    }

    var my_div1 = $("#dvd");
    if (response['dvd']) {
	my_div1.prop('checked', true);
    } else {
	my_div1.prop('checked', false);

    }

    var my_div1 = $("#blu_ray");
    if (response['blu_ray']) {
	my_div1.prop('checked', true);
    } else {
	my_div1.prop('checked', false);

    }

    var my_div1 = $("#doc_cam");
    if (response['doc_cam']) {
	my_div1.prop('checked', true);
    } else {
	my_div1.prop('checked', false);

    }

     var my_div1 = $("#extro");
    if (response['extro']) {
	my_div1.prop('checked', true);
    } else {
	my_div1.prop('checked', false);

    }

    var my_div1 = $("#audio");
    if (response['audio']) {
	my_div1.prop('checked', true);
    } else {
	my_div1.prop('checked', false);

    }

    var my_div1 = $("#mondopad");
    if (response['mondopad']) {
	my_div1.prop('checked', true);
    } else {
	my_div1.prop('checked', false);

    }

    var my_div1 = $("#tech_chart");
    if (response['tech_chart']) {
	my_div1.prop('checked', true);
    } else {
	my_div1.prop('checked', false);

    }

}

/*This function serves to take data from the python
file and dumps into html file on the UI after taking from the education_detail()*/
function seteducationTech() {
    setRoomId(getRoomId());
    if(getRoomId()){
         var url = '/getEducationData/'+ getRoomId();
        //  console.log(url)
         $.ajax({
                url: url,
                dataType: 'json',
                type:'GET',
                success: function(response){
                     if (response["success"] != 0) {//If successful
			                  education_detail(response); //a function with education tech details
                                                    }
                },
                error: function(error) {
                console.log(error);
                }
            });
    }
}

    /*this functions saves the edecuationtech materials on the front-end and updated them as thier values change*/
function saveEdTechChanges(roomID){
    var edtechDetails = {}//For passing into Ajax data field (multiple attributes to pass)
    edtechDetails["projector"] = document.getElementById('projectors').value;
    edtechDetails["smartboards"] = document.getElementById('smartboards').value;
    edtechDetails["instructor_computers"] = document.getElementById('instructor_computers').value;
    edtechDetails["podium"] = document.getElementById('podiums').value;
    edtechDetails["student_workspace"] = document.getElementById('student_workspace').value;
    edtechDetails["chalkboards"] = document.getElementById('chalkboards').value;
    edtechDetails["whiteboards"] = document.getElementById('whiteboards').value;
    edtechDetails["vhs"] = document.getElementById('vhs').checked;
    edtechDetails["dvd"] = document.getElementById('dvd').checked;
    edtechDetails["blu_ray"] = document.getElementById('blu_ray').checked;
    edtechDetails["audio"] = document.getElementById('audio').checked;
    edtechDetails["mondopad"] = document.getElementById('mondopad').checked;
    edtechDetails["doc_cam"] = document.getElementById('doc_cam').checked;
    edtechDetails["extro"] = document.getElementById('extro').checked;
    edtechDetails["tech_chart"] = document.getElementById('tech_chart').checked;
    var url = '/saveEdTechChanges/'+getRoomId();
         $.ajax({
             type: "POST",
                url: url,
                data: edtechDetails,
                dataType: 'json',
                success: function(response){
                    if (response["success"] != 0) {
                        console.log("SUCCESSFUL JS AJAX CALL")
                    }
                    else{

                        console.log("Else in ajax")
                    }
                },
                error: function(error){
                    console.log("ERROR")

                }
         });
}


/* keeps tracks of active modals, making sure that they remain responsive after closing an overlaying instance*/
var modal_counter = 0;
$(document).ready(function () {
  Dropzone.options.myId = false;

  $('.modal').on('shown.bs.modal', function () {
      modal_counter++;
  });
  $('.modal').on('hidden.bs.modal', function () {
      modal_counter--;
      if(modal_counter){
          $('body').addClass('modal-open');
      }
      else{
          $('body').removeClass('modal-open');
      }
  });
})

File: app/controllers/main_routes/buildingManagement.py
from app.controllers.main_routes import *
from app.controllers.main_routes.main_routes import *

from app.allImports import *
from flask import render_template,flash
from app.logic.redirectBack import redirect_url
import json
from app.logic import courseLogic
from app.logic import functions
from app.models.models import Building, Rooms, BuildingManager
from app.loadConfig import load_config
from datetime import datetime

from werkzeug.utils import secure_filename

#Running page
@main_bp.route("/buildingManagement", methods=["GET"])
def buildingManagement():                                   #Gathering of appropriate data to send to html
    au = AuthorizedUser()
    cfg = load_config()
    if au.user.isAdmin:
        building = Building.select()
    else:
        building = (Building.select()
                    .join(BuildingManager)
                    .where(BuildingManager.username == au.username)
                    )
    le_rooms = []                                           #Scott LITERALLY created this variable with a terrible name
    if len(building) == 0:
        abort(403)
    else:
        for b in building:
            rooms = Rooms.select().where(Rooms.building == b.bID)
            for room in rooms:
                le_rooms.append(room)
        user = User.select()
        return render_template("buildingManagement.html",
                                building = building,
                                rooms = le_rooms,
                                user = user,
                                cfg = cfg,
                                isAdmin = au.user.isAdmin
                                )

@main_bp.route('/getRoomData/<rID>', methods=["GET"])
# connected to ajax calls in the javascript file to populate the room data into panel
def getRoomData(rID):
    room = Rooms.get(Rooms.rID == rID)                          #Sets room variable to room object where the rID's are the same
    room_details={}                                             #Empty dictionare to hold all room attributes
    room_details["number"] = room.number                        #Begin setting room attributes to their appropriate keys
    room_details["building"] = room.building.shortName
    room_details["capacity"] = room.maxCapacity
    room_details["type"] = room.roomType
    room_details["specializedEq"] = room.specializedEq
    room_details["specialFeatures"] = room.specialFeatures
    room_details["movableFurniture"] = room.movableFurniture    #Use to make sure all above attributes are correct?
    room_details["visualAcc"] = room.visualAcc
    room_details["audioAcc"] = room.audioAcc
    room_details["physicalAcc"] = room.physicalAcc
    return json.dumps(room_details)

@main_bp.route("/imageUpload/<rid>", methods=["POST"])
def imageUpload(rid):
    print("RID: ", rid)
    # try:
        # Save the files
    file = request.files['file']
    filename = secure_filename(file.filename)
    file.save(os.path.join('app/static/images/', filename))
    # except Exception as e:
    #     print(e)
    #     return ("Image failed to upload to the server", 500)

    # Update the DB
    room = Rooms.get(Rooms.rID == rid)
    if not room.roomImageURL:
        room.roomImageURL = filename
    elif len(room.roomImageURL) == 0:
        room.roomImageURL = filename
    else:
        room.roomImageURL += "," + filename
    room.save()
    print("Room image saved", room.roomImageURL)
    return ("Success", 200)

@main_bp.route("/getImages/<rid>", methods=["GET"])
def getImages(rid):
    images = Rooms.get(Rooms.rID == rid)
    imageSet = []
    if not images.roomImageURL:
        return json.dumps(imageSet)
    if len(images.roomImageURL) == 0:
        return json.dumps(imageSet)
    for img in images.roomImageURL.split(","):
        try:
            fileSize = os.path.getsize("app/static/images/" + img);
            imageSet.append({"name": img, "size": fileSize})
        except:
            images.roomImageURL.replace(img, "")
            images.save()
    return json.dumps(imageSet)

@main_bp.route("/removeImage", methods=["POST"])
def removeImage():
    room = Rooms.get(Rooms.rID == request.form["rid"])
    if not len(room.roomImageURL) == 0:
        os.remove("app/static/images/" + room.roomImageURL)
        room.roomImageURL = None
    room.save()
    return json.dumps({"success":1})

@main_bp.route("/saveChanges/<rID>", methods=["POST"])
def saveChanges(rID):
#Saves all room data EXCLUDING ed tech
   try:
        room = Rooms.get(Rooms.rID==rID)                        #Sets room variable to room object where the rID's are the same
        data = request.form
        room.maxCapacity = (data['roomCapacity'])               #Begin setting room attributes to their keys in js (Inside saveChanges)
        room.roomType = (data['roomType'])
        room.specializedEq = (data['specializedEq'])
        room.specialFeatures = (data['specialFeatures'])
        if data['movableFurniture'] == 'false':                 #If movable furniture reads false from js
            room.movableFurniture = 0                           #Set to false in sqlite
        else:
            room.movableFurniture = 1                           #Else: its set to true in sqlite
        room.visualAcc = (data['visualAcc'])
        room.audioAcc = (data['audioAcc'])
        room.physicalAcc = (data['physicalAcc'])
        # e.g., 06/02/2020 11:28:36 PM
        room.lastModified = (datetime.now().strftime("%m/%d/%Y %I:%M %p"))
        room.save()                                             #Save data
        return json.dumps({"success":1})
   except:
       flash("An error has occurred, your changes were NOT saved. Please try again.","error")
       return json.dumps({"error":0})

#Education tech pull
@main_bp.route('/getEducationData/<rid>', methods = ["GET"])
def getEducationData(rid):
    room = Rooms.get(Rooms.rID == rid)
    tech_details = room.educationTech
    education_materials={}
    education_materials["projector"] = tech_details.projector
    education_materials["smartboards"] = tech_details.smartboards
    education_materials["instructor_computers"] = tech_details.instructor_computers
    education_materials["podium"] = tech_details.podium
    education_materials["student_workspace"] = tech_details.student_workspace
    education_materials["chalkboards"] = tech_details.chalkboards
    education_materials["whiteboards"] = tech_details.whiteboards
    education_materials["dvd"]=tech_details.dvd
    education_materials["blu_ray"]= tech_details.blu_ray
    education_materials["audio"]= tech_details.audio
    education_materials["extro"]=tech_details.extro
    education_materials["doc_cam"]=tech_details.doc_cam
    education_materials["vhs"]= tech_details.vhs
    education_materials["mondopad"]=tech_details.mondopad
    education_materials["tech_chart"]=tech_details.tech_chart
    return json.dumps(education_materials)

#Education tech save to DB
@main_bp.route("/saveEdTechChanges/<rid>", methods=["POST"])
def saveEdTechChanges(rid):
  try:
    room = Rooms.get(Rooms.rID==rid)
    edtech_update = room.educationTech
    data = request.form
    edtech_update.projector = data['projector']
    edtech_update.smartboards = data['smartboards']
    edtech_update.instructor_computers = data['instructor_computers']
    edtech_update.podium= data['podium']
    edtech_update.student_workspace = data['student_workspace']
    edtech_update.chalkboards= data['chalkboards']
    edtech_update.whiteboards= data['whiteboards']
    if data['vhs'] == 'false':                                  #this checks the booleans and set them to correct values.
                                                                #If they are checked thier values set to 1s an if they are unchecked thier values sets to 1s
        edtech_update.vhs = 0
    else:
        edtech_update.vhs = 1
    if data['dvd'] == 'false':
        edtech_update.dvd= 0
    else:
        edtech_update.dvd = 1
    if data['blu_ray']=='false':
        edtech_update.blu_ray=0
    else:
        edtech_update.blu_ray=1
    if data['audio']=='true':
        edtech_update.audio=1
    else:
        edtech_update.audio=0
    if data['mondopad']=='true':
        edtech_update.mondopad=1
    else:
        edtech_update.mondopad=0
    if data['doc_cam']=='true':
        edtech_update.doc_cam=1
    else:
        edtech_update.doc_cam=0
    if data['tech_chart']=='true':
        edtech_update.tech_chart=1
    else:
        edtech_update.tech_chart=0
    if data['extro']=='false':
        edtech_update.extro=0
    else:
        edtech_update.extro=1
    edtech_update.save()
    flash("Your changes have been saved!")
    return json.dumps({"success":1})
  except:
      flash("An error has occurred, your changes were NOT saved. Please try again.","error")
      return json.dumps({"error":0})
